engine/class_modules/sc_shaman.cpp

```diff
@@ -8536,6 +8635,26 @@ void shaman_t::create_buffs()
                             ->set_default_value_from_effect_type( A_ADD_PCT_MODIFIER, P_TICK_TIME )
                             ->apply_affecting_conduit( conduit.call_of_flame );
 
+  buff.fireheart = make_buff( this, "fireheart", spell.t28_2pc_ele )
+                       // TODO: confirm if duration is affected by conduit and tier set 4pc bonus
+                       ->set_duration( buff.fire_elemental->buff_duration() )
+                       ->set_tick_callback( [ this ]( buff_t* b, int, timespan_t ) {
+                         if ( buff.lava_surge->check() )
+                         {
+                           proc.wasted_lava_surge->occur();
+                         }
+
+                         proc.lava_surge->occur();
+                         if ( !sim->active_player->executing || sim->active_player->executing->id != 51505 )
```

Comment:
This will break badly in multi-actor sims. Use "source->executing" here.

Reply:
How do I access source? Where do I find that?

Reply:
The buff passed to the lambda has a source, but since you are capturing this, you could just use "executing", which would in fact be the shaman object's executing. That's probably for the best.

I'd just create a generic "trigger_lava_surge" method in the shaman module (like it has for many other kind of behavior), and all it from flame shock and this spot, since the behavior seems indentical?

Reply:
Hawk found it. Added.